packages/common/src/utils.ts

```diff
@@ -93,19 +93,39 @@ function parseGethParams(json: any) {
     [Hardfork.Berlin]: 'berlinBlock',
     [Hardfork.London]: 'londonBlock',
     [Hardfork.MergeForkIdTransition]: 'mergeForkBlock',
+    [Hardfork.Shanghai]: 'shanghaiBlock',
   }
-  params.hardforks = Object.values(Hardfork)
-    .map((name) => ({
-      name,
-      block: name === Hardfork.Chainstart ? 0 : config[forkMap[name]] ?? null,
+  const forkMapRev = Object.keys(forkMap).reduce((acc, elem) => {
+    acc[forkMap[elem]] = elem
+    return acc
+  }, {} as { [key: string]: string })
+  const configHardforks = Object.keys(config).filter((key) => forkMapRev[key] !== undefined)
+
+  params.hardforks = configHardforks
+    .map((nameBlock) => ({
+      name: forkMapRev[nameBlock],
+      block: config[nameBlock],
     }))
     .filter((fork) => fork.block !== null)
+  const mergeConfig = {
+    name: Hardfork.Merge,
+    ttd: config.terminalTotalDifficulty,
+    block: null,
+  }
+  params.hardforks.unshift({ name: Hardfork.Chainstart, block: 0 })
+  const nonzeroIndex = params.hardforks.findIndex((hf: any) => hf.block > 0)
+  if (
+    (config.terminalTotalDifficultyPassed === true ||
+      config.terminalTotalDifficultyPassed === 'true') &&
+    nonzeroIndex !== -1
+  ) {
+    // find index where block > 0
+    params.hardforks.splice(nonzeroIndex, 0, mergeConfig)
```

Comment:
ummm, doesn't actually matter in scheme of things if those are all at 0
the issue is when terminalTotalDifficultyPassed is false or not present, then where should we place Merge?

The order we assume is the order they showup in the `config` section, and it would have been nice to also have `mergeBlock` there even if its block is null

So that there is no confusion about the order, may be we can give this feedback to the geth guys and if the mergeBlock is not present in the config (like it doesn't for now) we can use this logic to place merge

Reply:
K, [this](https://github.com/ethereum/go-ethereum/pull/25769) is the PR where they add it.  I'm honestly not sure how to handle this one.  The main place they reference this new parameter is [here](https://github.com/ethereum/go-ethereum/blob/6fd06ab0751f1312d48bf593c7f356dcd6f51ade/eth/sync.go#L154) where the sync type is being decided (so akin to how we differentiate between full sync and beacon sync).  Can you do beaconsync forward before the merge block or does the client have to continue with full sync till it hits the merge block?  If the latter, I still think we should explicitly look for the london fork block number and insert this one after it.  I can't think of a reason we should have GrayGlacier or ArrowGlacier in a custom network so we shouldn't need to check for them.

Reply:
but the PoS or PoW consensus actually has nothing to do with pre/post london, it can be placed anywhere

Reply:
ok, looking at the geths PR has got me confused about this flag, but we defintely need post merge hardforks differentiation which certainly needs merge, like withdrawals, and 4844. I am not clear if thats the way to go especially with mergeForkID confusion of it bein pre/post merge. 
But since kiln, no network has had mergeForkId before merge so may be i am over thinking it?

Reply:
ok i have tried to make this thing a bit more intelligenct, placing merge before the first postmerge hardfork,  and is still dynamic with respect to mergeForkId block

Reply:
Why is Geth not just giving a block number for the Merge?

Reply:
All this dynamic handling and wild-guessing for this one time Merge event which has now passed seems totally overblown to me and doesn't really solve any problems any more (right?).

Reply:
correct, may be they haven't had yet added handling for merge block
Also the genesis file seems weird, so the genesis block has difficulty and extraData belonging to PoA consensus, but shanghai is also at 0 for withdrawals so obviously merge hf has passed

Reply:
@holgerd77 i think the latest implementation is logical:

place the merge hardfork before any merge dependent hardforks specified on non zero block as genesis can't be a PoS block.

Our dynamic handling of hardfork sequences and hardfork changes can handle the rest.